### PR TITLE
[WIP] [CMake] Add a mode SWIFT_INSTALL_AS_SYMLINK

### PR DESCRIPTION
**Pull request by jrose-apple**
Alternative approach to #13220 for generating an xctoolchain made of symlinks. I don't think I like it as much, but what do others think?

Pros:
- Piggybacks on CMake's notion of what gets installed, including component control, instead of having a separate list in an external script (build-script-impl)

Cons:
- Relies on `swift_install_in_component`, which ideally we'd be able to drop entirely in favor of CMake's `install` or at least LLVM's notion of components.
- Basically reverse-engineering the arguments of `install`, which is not great.
- Does not automatically pick up new libraries like #13220 would (though that's less important these days). #13220 symlinks all of lib/swift/; this has symlinks for everything *in* lib/swift/
- Probably has weird silent failures.
- No obvious way to make it work for LLDB and/or Clang in addition to Swift.

### CODE REVIEW

**Comment by jrose-apple** [comment]
Oops, I meant to mark this as a draft now that we can do that. Oh well, whatever.

**Comment by benlangmuir** [comment]
How much does this save vs. just doing a normal install?  My experience has been that copying is not the bottleneck, particularly if not much has changed since the last build.

**Comment by jrose-apple** [comment]
The point isn't to save time over one install; it's to not have to run the install at all between builds. Very convenient if you're using ninja directly to just recompile the compiler.

**Comment by benlangmuir** [comment]
So something like this:
* Build a toolchain either with `build-script --install*`, or using `build-toolchain` plus your new cmake flag `-DSWIFT_INSTALL_AS_SYMLINKS=1`.
  * This is a normal toolchain, *except* for the products that come from swift (e.g. clang or libdispatch is a copy, but swift is a symlink)
* Incremental rebuilds of swift are reflected immediately without re-installing.  E.g. you rebuild the compiler without rebuilding the stdlib.

Seems reasonable.

File: cmake/modules/SwiftComponents.cmake [hunk 1]
```diff
@@ -162,7 +162,27 @@ function(swift_install_in_component)
     return()
   endif()
 
-  install(${ARGN})
+  if((NOT SWIFT_INSTALL_AS_SYMLINKS) OR
+     (NOT ARG_DESTINATION) OR
+     (ARG_RENAME))
```
Comment (benlangmuir) [inline cmake/modules/SwiftComponents.cmake]: RENAME seems to only be used by test suite, but where does `NOT ARG_DESTINATION` occur in practice?

Comment (jrose-apple) [reply cmake/modules/SwiftComponents.cmake]: I don't think it ever occurs, but if someone didn't specify a destination and relied on CMake's default locations we'd need to add more logic here. I could make it fail loudly instead of falling back to a regular install if you think that's better.

**Comment by jrose-apple** [comment]
Opinions on this approach vs. the previous one?

File: cmake/modules/SwiftComponents.cmake (hunk 1 above)
Comment (benlangmuir) [reply cmake/modules/SwiftComponents.cmake]: Maybe a warning?

**Comment by benlangmuir** [comment]
> Opinions on this approach vs. the previous one?

I think building on cmake is much more robust.  This approach can work within a complete toolchain, e.g. including swiftpm and downstream packages like sourcekit-lsp. The previous approach was unlikely to ever handle that robustly.  I attempted to do something like that inside build-script-impl itself in order to build sourcekit-lsp, before abandoning that approach.  It's just not reasonable to hardcode dozens of paths inside the build directories of all our separate projects, especially on Linux where we need to get dispatch, foundation, and xctest.

**Comment by jrose-apple** [comment]
@compnerd, I tagged you specifically because of the need to interpose on `install`. What are your feelings about that? cc also @gottesmm

**Comment by compnerd** [comment]
I'm really not a fan of the inter-positioning.  The custom wrappers make it much more difficult for someone to come in to the project and start to be productive.  I think that we should actually just do the layout by setting the output directories instead (that is, have a `bin`, `lib` in the build tree, no need for symlinks at all).  It would mean that our toolchain is also actually relocatable that way, as the structure would be the same as the on-disk layout of the toolchain, just with `/usr` dropped.

**Comment by jrose-apple** [comment]
Just having bin/, lib/, etc. in the build tree isn't good enough because we want multiple projects' artifacts to be in the toolchain. (I just started with Swift because it's the most important.) Hm.

**Comment by futurejones** [comment]
Can one of the admins verify this patch?

**Comment by shahmishal** [comment]
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))
